File: Daily/PY/Leetcode331-验证二叉树的前序序列化.py
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        n = len(preorder)
        i = 0
        stk = [1]
        while i < n:
            if not stk:
                return False
            if preorder[i] == ',':
                i += 1
            elif preorder[i] == '#':
                stk[-1] -= 1
                if stk[-1] == 0:
                    stk.pop()
                i += 1
            else:
                while i < n and preorder[i] != ',':
                    i += 1
                stk[-1] -= 1
                if stk[-1] == 0:
                    stk.pop()
                stk.append(2)
        return not stk

File: Leetcode/sbgo/Leetcode331-验证二叉树的前序序列化.go
package sbgo

func isValidSerialization(preorder string) bool {
	n := len(preorder)
	stk := []int{1}
	i := 0
	for i < n {
		if len(stk) == 0 {
			return false
		}

		if preorder[i] == ',' {
			i++
		} else if preorder[i] == '#' {
			stk[len(stk)-1]--
			if stk[len(stk)-1] == 0 {
				stk = stk[:len(stk)-1]
			}
			i++
		} else {
			for i < n && preorder[i] != ',' {
				i++
			}
			stk[len(stk)-1]--
			if stk[len(stk)-1] == 0 {
				stk = stk[:len(stk)-1]
			}
			stk = append(stk, 2)
		}
	}
	return len(stk) == 0
}
